lib/msf/core/post/common.rb

```diff
@@ -247,5 +247,42 @@ def command_exists?(cmd)
     raise "Unable to check if command `#{cmd}' exists"
   end
 
+  # Executes +cmd+ on the remote system and return an array containing the
+  # output and if it was successful or not.
+  #
+  # This is simply a wrapper around {#cmd_exec} that also checks the exit code
+  # to determine if the execution was successful or not.
+  #
+  # @param [String] cmd The command to execute
+  # @param [String] arg The optional arguments of the command (can de included in +cmd+ instead)
+  # @param [Integer] timeout The time in sec. to wait before giving up
+  # @param [Hash] opts An Hash of options (see {#cmd_exec})
+  # @return [Array(String, Boolean)] Array containing the output string
+  #   followed by a boolean indicating if the command succeded or not
+  def cmd_exec_with_result(cmd, args = nil, timeout = 15, opts = {})
+    verification_token = Rex::Text.rand_text_alphanumeric(8)
+
+    _cmd = cmd.dup
+    _cmd << " #{args}" if args
+    if session.platform == 'windows'
+      if session.type == 'powershell'
+        # The & operator is reserved by Powershell and needs to be wrapped in double quotes
+        result = cmd_exec('cmd', "/c #{_cmd} \"&\" if not errorlevel 1 echo #{verification_token}", timeout, opts)
+      else
+        result = cmd_exec('cmd', "/c #{_cmd} & if not errorlevel 1 echo #{verification_token}", timeout, opts)
+      end
+    else
+      result = cmd_exec('command', "#{_cmd} && echo #{verification_token}", timeout, opts)
+    end
+
+    if result.include?(verification_token)
+      # Removing the verification token to cleanup the output string
+      [result.lines[0...-1].join.strip, true]
+    else
+      [result.strip, false]
+    end
+  rescue Rex::TimeoutError => e
+    [e.message, false]
```

Comment:
The `e.message` shouldn't be used here because it can't be differentiated from an OS command that printed output and exited with an error code.

I would suggest instead something like:
```suggestion
  rescue Rex::TimeoutError => e
    elog("Command timed out: #{e.message}")
    ['', false]
```

Reply:
I'm happy to change this, but I'm not sure to understand. If a `Rex::TimeoutError` exception is raised, the error message will be something like `Operation timed out`, which will give an explanation of why it failed (`false` is returned in the second element of the array). Returning an empty string will not inform the user if the operation failed because the command actually failed or if a timeout occurred.

Unless I'm missing something, the idea is to always return the command output when the second element is `true` (success) and always return an error message when it is `false` (failure), whether the error comes from the command `stderr` or from a timeout. Do you think about a specific use case where this assumption might be wrong?

Reply:
Addressed in [2bf0107](https://github.com/rapid7/metasploit-framework/pull/16488/commits/2bf010786511d447767dfa6b931aeaadc6d4216d)